File: cpp/_875.cpp
// O(NlogM) / O(1)
class Solution {
public:
    int minEatingSpeed(vector<int>& piles, int H) {
        int l = 1, r = *max_element(piles.begin(), piles.end());
        while (l < r) {
            int m = l + (r - l) / 2;
            if (count(piles, m) <= H)
                r = m;
            else
                l = m + 1;
        }
        return l;
    }
private:
    int count(vector<int>& piles, int k) {
        int ans = 0;
        for (auto& p : piles) {
            ans += (p - 1 + k) / k; // ceil
        }
        return ans;
    }
};

File: python/_875.py
# O(NlogM) / O(1)
class Solution:
    def minEatingSpeed(self, piles: List[int], H: int) -> int:
        
        def count(k):
            hours = 0
            for p in piles:
                hours += (p - 1 + k) // k # ceil
            return hours
        
        l, r = 1, max(piles)
        while l < r:
            m = l + (r - l) // 2
            if count(m) <= H:
                r = m
            else:
                l = m + 1
        return l